[BROKEN] Asynchronous handlers

szmarczak opened the pull request:
Fixes #701 

#### Checklist

- [x] I have read the documentation.
- [x] I have included a pull request description of my changes.
- [x] I have included some tests.

sindresorhus commented on source/as-promise.js:
```diff
@@ -2,110 +2,122 @@
+const asPromise = () => {
+	let options = {};
 	const proxy = new EventEmitter();
+	const {promise, resolve, reject, onCancel} = detachedPCancelable();
```
This pattern is called deferred promise and is generally an anti-pattern, as unlike the constructors, thrown errors won't be caught and won't result in promise rejection.

szmarczak replied:
> thrown errors won't be caught

Can you elaborate on this one? I'm not sure what do you mean by that...

sindresorhus commented on source/create.js:
```diff
@@ -27,16 +26,23 @@ const create = defaults => {
 		defaults.handler = (options, next) => next(options);
 	}
 
+	if (defaults.handler.constructor.name !== 'AsyncPromise') {
```
This is not a reliable check. What are you trying to do here?

szmarczak replied:
Nevermind. I'll fix it :P

sindresorhus commented on source/create.js:
```diff
@@ -27,16 +26,23 @@ const create = defaults => {
+		defaults.handler = util.promisify(defaults.handler);
+	}
+
 	function got(url, options) {
```
Use arrow function.

sindresorhus commented on source/create.js:
```diff
@@ -27,16 +26,23 @@ const create = defaults => {
+			});
+		}).catch(error => {
+			promiseOrStream._reject(error);
+		});
```
This could be simplified to:

```js
(async () => {
	try {
		const normalizedOptions = await normalizeArguments(url, options, defaults);
		await defaults.handler(normalizedOptions, promiseOrStream._sendRequest);
	} catch (error) {
		promiseOrStream._reject(error);
	}
})();
```

sindresorhus commented:
Can you also update the `got.create()` docs showing how the async handlers will be used?

szmarczak commented:
Reworked, it's much readable now :P The previous version of this PR was too bloated.

> Can you explain your changes?

In short, I added `await pEvent(proxy, 'options')` before `requestAsEventEmitter(options)`.
So normalizing options can be async. And if that can be async, then **all** hooks can be async too - including handlers ;)

> Can you also update the `got.create()` docs showing how the async handlers will be used?

What do you mean?
